### PR TITLE
Fix IPCWriter for Sliced BooleanArray

### LINKED ISSUE #3496: Sliced batch w/ bool column doesn't roundtrip through IPC
**Describe the bug**
It seems that the IPC writer doesn't correctly account for sliced boolean columns.

**To Reproduce**
```rust
#[test]
fn encode_bools() {
    let val_bool_field = Field::new("val", DataType::Boolean, false);

    let schema = Arc::new(Schema::new(vec![val_bool_field]));

    let bools = BooleanArray::from(vec![true, false]);

    let batch =
        RecordBatch::try_new(Arc::clone(&schema), vec![Arc::new(bools)]).unwrap();
    let batch = batch.slice(1, 1);

    let mut writer = StreamWriter::try_new(Vec::<u8>::new(), &schema).unwrap();
    writer.write(&batch).unwrap();
    writer.finish().unwrap();
    let data = writer.into_inner().unwrap();

    let mut reader = StreamReader::try_new(Cursor::new(data), None).unwrap();
    let batch2 = reader.next().unwrap().unwrap();
    assert_eq!(batch, batch2);
}
```

This fails with:

```text
thread 'writer::tests::encode_bools' panicked at 'assertion failed: `(left == right)`
  left: `RecordBatch { schema: Schema { fields: [Field { name: "val", data_type: Boolean, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }, columns: [BooleanArray
[
  false,
]], row_count: 1 }`,
 right: `RecordBatch { schema: Schema { fields: [Field { name: "val", data_type: Boolean, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }, columns: [BooleanArray
[
  true,
]], row_count: 1 }`', arrow-ipc/src/writer.rs:1953:9
```

**Expected behavior**
Test passes.

**Additional context**
Tested on commit https://github.com/apache/arrow-rs/commit/eae993fd196d0a8df8a90857bc4a7ae8f5a3e845 .

Originally observed in https://github.com/influxdata/influxdb_iox/issues/6515 . There we've also looked into the raw IPC data which lead to the conclusion that this is likely related to boolean columns (since other columns seem to work) and a writer issue (not a reader issue).

### PR DESCRIPTION
**Pull request by crepererum**
# Which issue does this PR close?
Closes #3496.

# Rationale for this change
It seems that the IPC writer doesn't correctly account for sliced boolean columns. See #3496.

# What changes are included in this PR?
New IPC writer routine for bools that shifts data correctly + tests.

# Are there any user-facing changes?
IPC for bools no longer broken.

### CODE REVIEW

File: arrow-ipc/src/writer.rs [hunk 1]
```diff
@@ -1231,6 +1231,38 @@ fn write_array_data(
                 compression_codec,
             )?;
         }
+    } else if matches!(data_type, DataType::Boolean) {
+        // Bools are special because the payload (= 1 bit) is smaller than the physical container elements (= bytes).
+        // The array data may not start at the physical boundary of the underlying buffer, so we need to shift bits around.
+        assert!(array_data.buffers().len() == 1);
+        let buffer = &array_data.buffers()[0];
```
Comment (tustvold) [inline arrow-ipc/src/writer.rs:1238]: Could we use Buffer::bit_slice here?

Comment (crepererum) [reply arrow-ipc/src/writer.rs]: Ah nice, didn't know that this existed. Fixed.

File: arrow-ipc/src/writer.rs [hunk 2]
```diff
@@ -1231,6 +1231,14 @@ fn write_array_data(
                 compression_codec,
             )?;
         }
+    } else if matches!(data_type, DataType::Boolean) {
+        // Bools are special because the payload (= 1 bit) is smaller than the physical container elements (= bytes).
+        // The array data may not start at the physical boundary of the underlying buffer, so we need to shift bits around.
+        assert!(array_data.buffers().len() == 1);
```
Comment (tustvold) [inline arrow-ipc/src/writer.rs:1237]: Assert_eq?

**Review by tustvold — APPROVED** [review]
Just a minor nit

File: arrow-ipc/src/writer.rs (hunk 2 above)
Comment (crepererum) [reply arrow-ipc/src/writer.rs]: done

**Pull request merged**